core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -498,104 +500,188 @@ class KafkaController(val config: KafkaConfig,
-   * 4. Wait until all replicas in RAR are in sync with the leader.
-   * 5  Move all replicas in RAR to OnlineReplica state.
-   * 6. Set AR to RAR in memory.
-   * 7. If the leader is not in RAR, elect a new leader from RAR. If new leader needs to be elected from RAR, a LeaderAndIsr
-   *    will be sent. If not, then leader epoch will be incremented in zookeeper and a LeaderAndIsr request will be sent.
-   *    In any case, the LeaderAndIsr request will have AR = RAR. This will prevent the leader from adding any replica in
-   *    RAR - OAR back in the isr.
-   * 8. Move all replicas in OAR - RAR to OfflineReplica state. As part of OfflineReplica state change, we shrink the
-   *    isr to remove OAR - RAR in zookeeper and send a LeaderAndIsr ONLY to the Leader to notify it of the shrunk isr.
-   *    After that, we send a StopReplica (delete = false) to the replicas in OAR - RAR.
-   * 9. Move all replicas in OAR - RAR to NonExistentReplica state. This will send a StopReplica (delete = true) to
-   *    the replicas in OAR - RAR to physically delete the replicas on disk.
-   * 10. Update AR in ZK with RAR.
-   * 11. Update the /admin/reassign_partitions path in ZK to remove this partition.
-   * 12. After electing leader, the replicas and isr information changes. So resend the update metadata request to every broker.
+   * Cleanup Phase: In the cases where this reassignment has to override an ongoing reassignment.
+   *   . The ongoing reassignment is in phase A
+   *   . ORS denotes the original replica set, prior to the ongoing reassignment
+   *   . URS denotes the unnecessary replicas, ones which are currently part of the ongoing reassignment but will not be part of the new one
+   *   . OVRS denotes the overlapping replica set - replicas which are part of the ongoing reassignment and will be part of the overriding reassignment
+   *       (it is essentially (RS - ORS) - URS)
+   *   1 Set RS = ORS + OVRS, AR = OVRS, RS = [] in memory
+   *   2 Send LeaderAndIsr request with RS = ORS + OVRS, AR = [], RS = [] to all brokers in ORS + OVRS
+   *     (because the ongoing reassignment is in phase A, we know we wouldn't have a leader in URS)
+   *   3 Replicas in URS -> Offline (force those replicas out of ISR)
+   *   4 Replicas in URS -> NonExistentReplica (force those replicas to be deleted)
+   *   5 Update ZK with RS = ORS + OVRS, AR = OVRS, RS = []
    *
-   * For example, if OAR = {1, 2, 3} and RAR = {4,5,6}, the values in the assigned replica (AR) and leader/isr path in ZK
-   * may go through the following transition.
-   * AR                 leader/isr
-   * {1,2,3}            1/{1,2,3}           (initial state)
-   * {1,2,3,4,5,6}      1/{1,2,3}           (step 2)
-   * {1,2,3,4,5,6}      1/{1,2,3,4,5,6}     (step 4)
-   * {1,2,3,4,5,6}      4/{1,2,3,4,5,6}     (step 7)
-   * {1,2,3,4,5,6}      4/{4,5,6}           (step 8)
-   * {4,5,6}            4/{4,5,6}           (step 10)
+   * Phase A: Initial trigger (when RAR != ISR)
```

Comment:
What is "RAR"?  Is this supposed to be TRS?

Reply:
Yes, good catch

Reply:
There are still a few cases where we're referring to RAR rather than TRS